AY2021S2-CS2103-T14-1/tp
# Fix(Parser): Update parser to deal with preamble better

### What this does:
- Fixes #166 
- Fixes #167 
- Fixes #199
- Fixes #293 

The parser now deals with non-integer and integer input separately. This
allows for better error messages to be raised to the user to explain why
the entered preamble used in index values may be failing.

This PR also separates non-integer input to throw an
illegalargumentexception instead of a parseexception, which is now
reserved for integer input that fails (<=0, > range of list, etc).

Additionally, this also fixes #388. This fix is done here as the master branch does not allow this specific fix to be tested properly.

### How to test
Run index-based commands (edit, delete) and ensure the behaviour is as follows:
1. Command without parameters should show command usage [ `oedit` ] should show "invalid command format [how to use]"
2. Command without non-integer index values should show command usage [ `oedit asdfa`, `oedit  1 asdfa`, `oedit 2s/dfa` ] should show "invalid command format [how to use]"
3. Command with integer <= 0 [ `odel -4` ] should show "Index is not a non-zero unsigned integer" 
4. Command with integer > valid range [ `odel 200` ] should show "The room index provided is invalid"
5. Check that all the edit commands specify that one of the optional params need to be specified.

Please run anything else that might break it or not perform in a user-friendly way to see if it still works.
